lib/min_heap.rb

```diff
@@ -58,14 +70,56 @@ def empty?
   # Time complexity: ?
   # Space complexity: ?
   def heap_up(index)
+    node = @store[index]
+    parent_index = (index - 1)/2
+    parent_node = @store[parent_index]
+    return @store if parent_node.key <= node.key || index == 0
+
+    @store[index] = parent_node
+    @store[parent_index] = node
     
+    index = parent_index
+    return heap_up(index)
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    parent_node = @store[index]
+    left_child_index = index * 2 + 1
+    right_child_index = index * 2 + 2
+    left_child = @store[left_child_index] 
+    right_child = @store[right_child_index]
+    if left_child.nil? == false && right_child.nil? == false
+      return @store if (parent_node.key <= left_child.key && parent_node.key <= right_child.key)
+
+      if left_child.key < right_child.key
+        @store[index] = left_child
+        @store[left_child_index] = parent_node
+        index = left_child_index
+        return heap_down(index)
+      else
+        @store[index] = right_child
+        @store[right_child_index] = parent_node
+        index = right_child_index
+        return heap_down(index)
+      end
+    elsif left_child.nil? == true && right_child.nil? == true
+      return @store 
+    elsif left_child.nil? == true && right_child.nil? == false
```

Comment:
I don't think the right child can't be `nil` if the left child is.

similarly I don't think the left child can be nil and the right not nil.